astrolabe/runner.py

```diff
@@ -139,59 +144,91 @@ def run(self, persist_cluster=False, startup_time=1):
-                self.cluster_name))
-            connection_string = self.get_connection_string()
-            load_test_data(connection_string, self.spec.driverWorkload)
-            LOGGER.info("Successfully loaded test data on cluster {!r}".format(
-                self.cluster_name))
-
         # Step-2: run driver workload.
         self.workload_runner.spawn(
             workload_executor=self.config.workload_executor,
             connection_string=self.get_connection_string(),
             driver_workload=self.spec.driverWorkload,
             startup_time=startup_time)
 
-        # Step-3: begin maintenance routine.
-        final_config = self.spec.maintenancePlan.final
-        cluster_config = final_config.clusterConfiguration
-        process_args = final_config.processArgs
-
-        if not cluster_config and not process_args:
-            raise RuntimeError("invalid maintenance plan")
-
-        if cluster_config:
-            LOGGER.info("Pushing cluster configuration update")
-            self.cluster_url.patch(**cluster_config)
-
-        if process_args:
-            LOGGER.info("Pushing process arguments update")
-            self.cluster_url.processArgs.patch(**process_args)
-
-        # Sleep before polling to give Atlas time to update cluster.stateName.
-        sleep(3)
-
-        # Step-4: wait until maintenance completes (cluster is IDLE).
-        selector = BooleanCallablePoller(
-            frequency=self.config.polling_frequency,
-            timeout=self.config.polling_timeout)
-        LOGGER.info("Waiting for cluster maintenance to complete")
-        selector.poll([self], attribute="is_cluster_state", args=("IDLE",),
-                      kwargs={})
-        self.verify_cluster_configuration_matches("final")
-        LOGGER.info("Cluster maintenance complete")
-
+        for operation in self.spec.operations:
+            if len(operation) != 1:
+                raise ValueError("Operation must have exactly one key: %s" % operation)
+                
+            op_name, op_spec = list(operation.items())[0]
+            
+            if op_name == 'setClusterConfiguration':
+                # Step-3: begin maintenance routine.
+                final_config = op_spec
+                cluster_config = final_config.clusterConfiguration
+                process_args = final_config.processArgs
+
+                if not cluster_config and not process_args:
+                    raise RuntimeError("invalid maintenance plan")
+
+                if cluster_config:
+                    LOGGER.info("Pushing cluster configuration update")
+                    self.cluster_url.patch(**cluster_config)
+
+                if process_args:
+                    LOGGER.info("Pushing process arguments update")
+                    self.cluster_url.processArgs.patch(**process_args)
+
+                # Step-4: wait until maintenance completes (cluster is IDLE).
+                self.wait_for_idle()
```

Comment:
I think you need to sleep for a couple of seconds before calling `waitForIdle` otherwise there could be situations in which you end up seeing the IDLE server state prematurely because maintenance has not yet been kicked off by the automation agent.

Reply:
I reported this for failover tests in https://jira.mongodb.org/browse/PRODTRIAGE-1232 (and the users are currently meant to add the sleeps in their scenarios to compensate), are you saying other actions are also affected?

Reply:
Yes. AFAIK none of the operations on clusters are applied synchronously. Operations are put in a task queue that is consumed by the automation agent. Consequently, this could cause strange failures even for other maintenance scenarios like cluster resizing. Adding a 2-3 second wait before waiting for idle should work.